File: backend/radar-service/src/main/java/com/radovan/spring/services/impl/PrometheusServiceImpl.java
package com.radovan.spring.services.impl;

import com.radovan.spring.services.PrometheusService;
import io.micrometer.core.instrument.*;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.OperatingSystemMXBean;
import java.lang.management.ThreadMXBean;
import java.util.concurrent.TimeUnit;

@Service
public class PrometheusServiceImpl implements PrometheusService {

	@Autowired
	private MeterRegistry registry;

	// 📈 Counters and Timers
	private Counter requestCounter;
	private Counter dbQueryCounter;
	private Counter externalApiLatencyCounter;
	private Timer responseTimer;

	// 📊 Gauges
	private Gauge heapUsageGauge;
	private Gauge heapAllocationRateGauge;
	private Gauge activeThreadsGauge;
	private Gauge cpuLoadGauge;
	private Gauge activeSessionsGauge;

	// 🔧 System beans
	private final MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
	private final ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
	private final OperatingSystemMXBean osMXBean = ManagementFactory.getOperatingSystemMXBean();

	@PostConstruct
	public void init() {
		// 📈 Counters
		requestCounter = registry.counter("api_requests_total");
		dbQueryCounter = registry.counter("db_query_total");
		externalApiLatencyCounter = registry.counter("external_api_latency_seconds");
		responseTimer = registry.timer("response_time_seconds");

		// 📊 Gauges (registered via lambda suppliers)
		Gauge.builder("heap_used_bytes", memoryMXBean, bean -> (double) bean.getHeapMemoryUsage().getUsed())
				.register(registry);

		Gauge.builder("heap_allocation_rate", memoryMXBean, bean -> (double) bean.getHeapMemoryUsage().getCommitted())
				.register(registry);

		Gauge.builder("active_threads_total", threadMXBean, bean -> (double) bean.getThreadCount()).register(registry);

		Gauge.builder("cpu_load_percentage", osMXBean, bean -> bean.getSystemLoadAverage()).register(registry);

		Gauge.builder("active_sessions", threadMXBean, bean -> (double) bean.getPeakThreadCount()).register(registry);
	}

	// 🔁 Metric update methods

	@Override
	public void increaseRequestCount() {
		requestCounter.increment();
	}

	@Override
	public void recordResponseTime(double duration) {
		responseTimer.record((long) (duration * 1000), TimeUnit.MILLISECONDS);
	}

	@Override
	public void updateMemoryUsage() {
		// Gauge auto-updates via supplier; no manual trigger needed
	}

	@Override
	public void updateThreadCount() {
		// Gauge auto-updates via supplier; no manual trigger needed
	}

	@Override
	public void updateCpuLoad() {
		// Gauge auto-updates via supplier; no manual trigger needed
	}

	@Override
	public void updateDatabaseQueryCount() {
		dbQueryCounter.increment();
	}

	@Override
	public void updateHeapAllocationRate() {
		// Gauge auto-updates via supplier; no manual trigger needed
	}

	@Override
	public void updateActiveSessions() {
		// Gauge auto-updates via supplier; no manual trigger needed
	}

	@Override
	public void updateHttpStatusCount(int statusCode) {
		int statusClass = statusCode / 100;
		switch (statusClass) {
		case 2 -> registry.counter("http_requests_status_total", "status", "2xx").increment();
		case 4 -> registry.counter("http_requests_status_total", "status", "4xx").increment();
		case 5 -> registry.counter("http_requests_status_total", "status", "5xx").increment();
		default -> {
			// Ignore 1xx, 3xx, etc.
		}
		}
	}

	@Override
	public void updateExternalApiLatency(double duration) {
		externalApiLatencyCounter.increment(duration);
	}

	@Override
	public void updateControllerMethodTags(String controller, String method) {
		registry.counter("http_requests_by_controller_method", "controller", controller, "method", method).increment();
	}
}

File: backend/api-gateway/src/main/scala/com/radovan/scalatra/services/impl/PrometheusServiceImpl.scala
package com.radovan.scalatra.services.impl

import com.radovan.scalatra.services.PrometheusService
import io.micrometer.core.instrument.{Counter, Gauge, MeterRegistry, Timer}
import jakarta.inject.{Inject, Singleton}

import java.lang.management.{ManagementFactory, MemoryMXBean, OperatingSystemMXBean, ThreadMXBean}
import java.util.concurrent.TimeUnit

@Singleton
class PrometheusServiceImpl @Inject() (registry: MeterRegistry) extends PrometheusService {

  private val requestCounter: Counter = registry.counter("api_requests_total")
  private val dbQueryCounter: Counter = registry.counter("db_query_total")
  private val httpStatusCounter200: Counter = registry.counter("http_requests_status_total", "status", "200")
  private val httpStatusCounter400: Counter = registry.counter("http_requests_status_total", "status", "400")
  private val httpStatusCounter500: Counter = registry.counter("http_requests_status_total", "status", "500")
  private val externalApiLatencyCounter: Counter = registry.counter("external_api_latency_seconds")
  private val responseTimer: Timer = registry.timer("response_time_seconds")

  private val memoryMXBean: MemoryMXBean = ManagementFactory.getMemoryMXBean
  private val threadMXBean: ThreadMXBean = ManagementFactory.getThreadMXBean
  private val osMXBean: OperatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean

  import java.util.function.ToDoubleFunction

  private val heapUsageGauge: Gauge = Gauge
    .builder("heap_used_bytes", memoryMXBean, new ToDoubleFunction[MemoryMXBean] {
      override def applyAsDouble(bean: MemoryMXBean): Double =
        bean.getHeapMemoryUsage.getUsed.toDouble
    })
    .register(registry)

  private val heapAllocationRateGauge: Gauge = Gauge
    .builder("heap_allocation_rate", memoryMXBean, new ToDoubleFunction[MemoryMXBean] {
      override def applyAsDouble(bean: MemoryMXBean): Double =
        bean.getHeapMemoryUsage.getCommitted.toDouble
    })
    .register(registry)

  private val activeThreadsGauge: Gauge = Gauge
    .builder("active_threads_total", threadMXBean, new ToDoubleFunction[ThreadMXBean] {
      override def applyAsDouble(bean: ThreadMXBean): Double =
        bean.getThreadCount.toDouble
    })
    .register(registry)

  private val cpuLoadGauge: Gauge = Gauge
    .builder("cpu_load_percentage", osMXBean, new ToDoubleFunction[OperatingSystemMXBean] {
      override def applyAsDouble(bean: OperatingSystemMXBean): Double =
        bean.getSystemLoadAverage
    })
    .register(registry)

  private val activeSessionsGauge: Gauge = Gauge
    .builder("active_sessions", threadMXBean, new ToDoubleFunction[ThreadMXBean] {
      override def applyAsDouble(bean: ThreadMXBean): Double =
        bean.getPeakThreadCount.toDouble
    })
    .register(registry)


  override def increaseRequestCount(): Unit = requestCounter.increment()

  override def recordResponseTime(duration: Double): Unit =
    responseTimer.record((duration * 1000).toLong, TimeUnit.MILLISECONDS)

  override def updateMemoryUsage(): Unit = heapUsageGauge.value()

  override def updateThreadCount(): Unit = activeThreadsGauge.value()

  override def updateCpuLoad(): Unit = cpuLoadGauge.value()

  override def updateDatabaseQueryCount(): Unit = dbQueryCounter.increment()

  override def updateHeapAllocationRate(): Unit = heapAllocationRateGauge.value()

  override def updateActiveSessions(): Unit = activeSessionsGauge.value()

  override def updateHttpStatusCount(statusCode: Int): Unit = {
    val statusClass = statusCode / 100
    statusClass match {
      case 2 => registry.counter("http_requests_status_total", "status", "2xx").increment()
      case 4 => registry.counter("http_requests_status_total", "status", "4xx").increment()
      case 5 => registry.counter("http_requests_status_total", "status", "5xx").increment()
      case _ => () // ignoriši ostale (1xx, 3xx, itd.)
    }
  }


  override def updateExternalApiLatency(duration: Double): Unit =
    externalApiLatencyCounter.increment(duration)
}